core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -57,6 +58,77 @@ trait WorkspaceComponent {
     def * = (namespace, name, id, bucketName, workflowCollection, createdDate, lastModified, createdBy, isLocked, recordVersion) <> (WorkspaceRecord.tupled, WorkspaceRecord.unapply)
   }
 
+  /** raw/optimized SQL queries for working with workspace attributes
+    */
+  object workspaceAttributesRawSqlQuery extends RawSqlQuery {
+    val driver: JdbcProfile = WorkspaceComponent.this.driver
+
+    // convenience case class to simplify signatures
+    case class WorkspaceAttributeWithReference(attribute: WorkspaceAttributeRecord,
+                                               entityRec: Option[EntityRecord])
+
+    // tells slick how to convert a result row from a raw sql query to an instance of WorkspaceAttributeWithReference
+    implicit val getAttributeWithReferenceResult:GetResult[WorkspaceAttributeWithReference] = GetResult { r =>
+      // note that the number and order of all the r.<< match precisely with the select clause of workspaceAttributesWithReferences
+      val attrRec: WorkspaceAttributeRecord = WorkspaceAttributeRecord(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<)
+
+      val entityIdOption: Option[Long] = r.<<
+      val entityRecOption = entityIdOption.map(id => EntityRecord(id, r.<<, r.<<, r.<<, r.<<, r.<<, r.<<))
+
+      WorkspaceAttributeWithReference(attrRec, entityRecOption)
+    }
+
+    /** direct replacement for WorkspaceComponent.workspaceAttributesWithReferences, now with optimized SQL
+      *
+      * @param workspaceIds the workspaces for which to query attributes
+      * @param attributeSpecs which attributes to return
+      * @return attributes found in the db, with optional entity values
+      */
+    def workspaceAttributesWithReferences(workspaceIds: Seq[UUID], attributeSpecs: WorkspaceAttributeSpecs): ReadAction[Seq[( (UUID, WorkspaceAttributeRecord), Option[EntityRecord] )]] = {
+      // if workspaceIds is empty, or the caller explicitly specified no attributes, short-circuit and return nothing
+      if (workspaceIds.isEmpty || (!attributeSpecs.all && attributeSpecs.attrsToSelect.isEmpty)) {
+        DBIO.successful(Seq())
+      } else {
+        // we have workspaceIds, and we will be retrieving attributes.
+        val workspaceIdList = reduceSqlActionsWithDelim(workspaceIds.map { id => sql"$id" })
+
+        // does this need "AND WORKSPACE_ATTRIBUTE.deleted = FALSE"?
+        // the method this replaced did not specify that, so we also don't specify it here.
```

Comment:
this is surprising. i'm trying to remember if this was intentional or not.

regardless, since you've left it equivalent to the current implementation, i don't see the problem with leaving it this way.

Reply:
(i'll let you know if I can remember the answer to this question and we can file a ticket if need be)